core_v2/sources/domains.move

```diff
@@ -505,14 +527,124 @@ module aptos_names_v2::domains {
         object::transfer_with_ref(
             object::generate_linear_transfer_ref(&record.transfer_ref),
             get_token_signer_address()
         );
         record.target_address = option::none();
     }
 
+    public fun renew_domain(
+        sign: &signer,
+        domain_name: String,
+        num_years: u8,
+    ) acquires CollectionCapabilityV2, NameRecord, RegisterNameEventsV1 {
+        // check the domain eligibility
+        let (is_valid, length) = utf8_utils::string_is_allowed(&domain_name);
+        assert!(is_valid, error::invalid_argument(EDOMAIN_HAS_INVALID_CHARACTERS));
+        assert!(length <= config::max_domain_length(), error::out_of_range(EDOMAIN_TOO_LONG));
+        assert!(length >= config::min_domain_length(), error::out_of_range(EDOMAIN_TOO_SHORT));
+
+        let price = price_model::price_for_domain_v1(length, num_years);
+        renew_domain_internal(sign, domain_name, num_years, price);
+    }
+
+    fun renew_domain_internal(
+        sign: &signer,
+        domain_name: String,
+        num_years: u8,
+        price: u64,
+    ) acquires CollectionCapabilityV2, NameRecord, RegisterNameEventsV1 {
+        validate_domain_to_renew(domain_name);
+
+        let record = get_record_mut(domain_name, option::none());
+        let registration_duration_secs = time_helper::years_to_seconds((num_years as u64));
+        record.expiration_time_sec = timestamp::now_seconds() + registration_duration_secs;
+
+        // pay the price
+        coin::transfer<AptosCoin>(sign, config::fund_destination_address(), price);
+        // log the event
+        event::emit_event<RegisterNameEventV1>(
+            &mut borrow_global_mut<RegisterNameEventsV1>(@aptos_names_v2).register_name_events,
+            RegisterNameEventV1 {
+                domain_name,
+                subdomain_name: option::none(),
+                registration_fee_octas: price,
+                expiration_time_secs: record.expiration_time_sec,
+                event_type: RENEW_NAME_EVENT,
+            },
+        );
+    }
+
+    fun validate_domain_to_renew(
+        domain_name: String,
+    ) acquires CollectionCapabilityV2, NameRecord {
+        // check if the domain is registered
+        assert!(name_is_registered(option::none(), domain_name), error::not_found(ENAME_NOT_EXIST));
+        // check if the domain is expired already
+        assert!(!name_is_expired(option::none(), domain_name), error::invalid_state(ENAME_NOT_AVAILABLE));
+        let record = get_record_mut(domain_name, option::none());
+        // check if the domain is expired in 6 month
+        let six_monthes  = 6 * 30 * 24 * 60 * 60;
+        // renew the domain
+        if (record.expiration_time_sec - timestamp::now_seconds() > six_monthes) {
+            assert!(false, error::invalid_state(EDOMAIN_NOT_AVAILABLE_TO_RENEW));
+        };
+    }
+
+    /// this is for domain owner to update subdomain expiration time
```

Comment:
I don't think we should leave that for subdomain owners if our principle is to leave domain owner full control of subdomains. but I'll still rename in case :P